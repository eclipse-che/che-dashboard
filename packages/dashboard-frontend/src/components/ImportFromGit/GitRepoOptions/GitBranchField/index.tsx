/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, TextInput } from '@patternfly/react-core';
import React from 'react';

export type Props = {
  onChange: (definition: string | undefined) => void;
  gitBranch: string | undefined;
};
export type State = {
  gitBranch: string | undefined;
};

export class GitBranchField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      gitBranch: this.props.gitBranch,
    };
  }

  public shouldComponentUpdate(nextProps: Readonly<Props>, nextState: Readonly<State>): boolean {
    return (
      this.state.gitBranch !== nextState.gitBranch || this.props.gitBranch !== nextProps.gitBranch
    );
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { gitBranch } = this.props;
    if (prevProps.gitBranch !== gitBranch) {
      this.setState({ gitBranch });
    }
  }

  private handleChange(value: string) {
    let gitBranch: string | undefined = value.trim();
    gitBranch = gitBranch !== '' ? gitBranch : undefined;
    if (gitBranch !== this.state.gitBranch) {
      this.setState({ gitBranch: value });
      this.props.onChange(value);
    }
  }

  public render() {
    const gitBranch = this.state.gitBranch || '';

    return (
      <FormGroup label="Git Branch">
        <TextInput
          aria-label="Git Branch"
          placeholder="Enter the branch of the Git Repository"
          onChange={value => this.handleChange(value)}
          value={gitBranch}
        />
      </FormGroup>
    );
  }
}
