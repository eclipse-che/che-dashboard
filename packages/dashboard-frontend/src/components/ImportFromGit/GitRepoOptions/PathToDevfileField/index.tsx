/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, TextInput } from '@patternfly/react-core';
import React from 'react';

export type Props = {
  onChange: (devfilePath: string | undefined) => void;
  devfilePath: string | undefined;
};
export type State = {
  devfilePath: string | undefined;
};

export class PathToDevfileField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      devfilePath: this.props.devfilePath,
    };
  }

  public shouldComponentUpdate(nextProps: Readonly<Props>, nextState: Readonly<State>): boolean {
    return (
      this.state.devfilePath !== nextState.devfilePath ||
      this.props.devfilePath !== nextProps.devfilePath
    );
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { devfilePath } = this.props;
    if (prevProps.devfilePath !== devfilePath) {
      this.setState({ devfilePath });
    }
  }

  private handleChange(value: string) {
    let devfilePath: string | undefined = value.trim();
    devfilePath = devfilePath !== '' ? devfilePath : undefined;
    if (devfilePath !== this.state.devfilePath) {
      this.setState({ devfilePath: value });
      this.props.onChange(value);
    }
  }

  public render() {
    const devfilePath = this.state.devfilePath || '';

    return (
      <FormGroup label="Path to Devfile">
        <TextInput
          aria-label="Path to Devfile"
          placeholder="Enter the relative path to the Devfile in the Git Repository"
          onChange={value => this.handleChange(value)}
          value={devfilePath}
        />
      </FormGroup>
    );
  }
}
