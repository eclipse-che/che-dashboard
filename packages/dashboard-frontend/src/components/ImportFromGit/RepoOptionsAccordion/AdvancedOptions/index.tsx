/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import React from 'react';

import { ContainerImageField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/ContainerImageField';
import { CpuLimitField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/CpuLimitField';
import { CreateNewIfExistingField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/CreateNewIfExistingField';
import { MemoryLimitField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/MemoryLimitField';
import { TemporaryStorageField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/TemporaryStorageField';

export type Props = {
  containerImage: string | undefined;
  temporaryStorage: boolean | undefined;
  createNewIfExisting: boolean | undefined;
  memoryLimit: number | undefined;
  cpuLimit: number | undefined;
  onChange: (
    containerImage: string | undefined,
    temporaryStorage: boolean | undefined,
    createNewIfExisting: boolean | undefined,
    memoryLimit: number | undefined,
    cpuLimit: number | undefined,
  ) => void;
};

export type State = {
  containerImage: string | undefined;
  temporaryStorage: boolean | undefined;
  createNewIfExisting: boolean | undefined;
  memoryLimit: number | undefined;
  cpuLimit: number | undefined;
};

export class AdvancedOptions extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      containerImage: props.containerImage,
      temporaryStorage: props.temporaryStorage,
      createNewIfExisting: props.createNewIfExisting,
      memoryLimit: props.memoryLimit,
      cpuLimit: props.cpuLimit,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { containerImage, temporaryStorage, createNewIfExisting, memoryLimit, cpuLimit } =
      this.props;

    if (containerImage !== prevProps.containerImage) {
      this.setState({ containerImage });
    }

    if (temporaryStorage !== prevProps.temporaryStorage) {
      this.setState({ temporaryStorage });
    }

    if (createNewIfExisting !== prevProps.createNewIfExisting) {
      this.setState({ createNewIfExisting });
    }

    if (memoryLimit !== prevProps.memoryLimit) {
      this.setState({ memoryLimit });
    }

    if (cpuLimit !== prevProps.cpuLimit) {
      this.setState({ cpuLimit });
    }
  }

  private handleContainerImage(containerImage: string | undefined) {
    const { temporaryStorage, createNewIfExisting, memoryLimit, cpuLimit } = this.state;

    this.setState({ containerImage });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  private handleTemporaryStorage(temporaryStorage: boolean | undefined) {
    const { containerImage, createNewIfExisting, memoryLimit, cpuLimit } = this.state;

    this.setState({ temporaryStorage });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  private handleCreateNewIfExisting(createNewIfExisting: boolean | undefined) {
    const { containerImage, temporaryStorage, memoryLimit, cpuLimit } = this.state;

    this.setState({ createNewIfExisting });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  private handleMemoryLimit(memoryLimit: number | undefined) {
    const { containerImage, temporaryStorage, createNewIfExisting, cpuLimit } = this.state;

    this.setState({ memoryLimit });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  private handleCpuLimit(cpuLimit: number | undefined) {
    const { containerImage, temporaryStorage, createNewIfExisting, memoryLimit } = this.state;

    this.setState({ cpuLimit });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  public render() {
    const { containerImage, temporaryStorage, createNewIfExisting, memoryLimit, cpuLimit } =
      this.state;
    return (
      <Form isHorizontal={true} onSubmit={e => e.preventDefault()}>
        <ContainerImageField
          onChange={containerImage => this.handleContainerImage(containerImage)}
          containerImage={containerImage}
        />
        <TemporaryStorageField
          onChange={temporaryStorage => this.handleTemporaryStorage(temporaryStorage)}
          isTemporary={temporaryStorage}
        />
        <CreateNewIfExistingField
          onChange={createNewIfExisting => this.handleCreateNewIfExisting(createNewIfExisting)}
          createNewIfExisting={createNewIfExisting}
        />
        <MemoryLimitField
          onChange={memoryLimit => this.handleMemoryLimit(memoryLimit)}
          memoryLimit={memoryLimit || 0}
        />
        <CpuLimitField
          onChange={cpuLimit => this.handleCpuLimit(cpuLimit)}
          cpuLimit={cpuLimit || 0}
        />
      </Form>
    );
  }
}
