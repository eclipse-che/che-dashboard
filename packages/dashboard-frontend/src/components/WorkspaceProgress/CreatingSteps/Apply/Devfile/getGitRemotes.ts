/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { V230DevfileProjects, V230DevfileProjectsItemsGit } from '@devfile/api';
import common from '@eclipse-che/common';

import { getProjectFromLocation } from '@/components/WorkspaceProgress/CreatingSteps/Apply/Devfile/getProjectFromLocation';
import devfileApi from '@/services/devfileApi';
import { getProjectName } from '@/services/helpers/getProjectName';

export interface GitRemote {
  name: string;
  url: string;
}

/**
 * Returns parsed Git remotes given a string in one of the following
 * formats:
 *
 * 1. Explicit name and URL
 * {{origin,https://git...},{upstream,https://git...},...}
 *
 * 2. URLs only, name is implicit: first is origin, second is upstream, subsequent are fork1, fork2
 * {https://git...,https://git...,...}
 *
 * @param remotes input string to parse
 * @returns parsed array of Git remotes
 */
export function getGitRemotes(remotes: string | undefined): GitRemote[] {
  if (remotes === undefined || remotes.length === 0) {
    return [];
  }
  const remotesArray = parseRemotes(remotes);
  if (Array.isArray(remotesArray[0])) {
    return parseNameAndUrls(remotesArray, remotes);
  }
  return parseUrls(remotesArray, remotes);
}

export function gitRemotesToParam(remotes: GitRemote[]): string {
  if (remotes.length === 0) {
    return '';
  }
  const remotesStr = `{${(remotes || [])
    ?.filter(remote => remote.url !== '')
    .map(remote => {
      const sortedRemoteKeys = Object.keys(remote).sort();
      const sortedRemoteObject = sortedRemoteKeys.reduce((acc, key) => {
        acc[key] = remote[key];
        return acc;
      }, {});
      return JSON.stringify(sortedRemoteObject);
    })
    .join(',')
    .replace(/"name":/g, '')
    .replace(/"url":/g, '')
    .replace(/"/g, '')}}`;

  return remotesStr;
}

function parseRemotes(remotes: string): string[] {
  try {
    return JSON.parse(sanitizeValue(remotes));
  } catch (e) {
    throw `Unable to parse remotes '${remotes}'. ${common.helpers.errors.getMessage(e)}`;
  }
}

/**
 * Replaces braces ({}) with square brackets ([]) and adds
 * quotation marks (") around strings
 * @param str
 * @returns string representing an array
 */
export function sanitizeValue(str: string): string {
  return (
    str
      .replace(/\s/g, '')
      .replace(/{/g, '[')
      .replace(/}/g, ']')
      /* eslint-disable no-useless-escape */
      // replace '[<string>' with '["<char>'
      .replace(/(\[([^\["]))/g, '["$2')
      // replace '<string>]' with '<char>"]'
      .replace(/(([^\]"])\])/g, '$2"]')
      // replace '<string>,<string>' with '<char>","<char>'
      .replace(/(([^\]"]),([^\["]))/g, '$2","$3')
    /* eslint-enable no-useless-escape */
  );
}

function parseNameAndUrls(remotesArray, remotesString): GitRemote[] {
  const remotes: GitRemote[] = [];
  remotesArray.forEach(value => {
    if (!Array.isArray(value) || value.length !== 2) {
      throw `Malformed remotes provided: ${remotesString}`;
    }
    remotes.push({ name: value[0], url: value[1] });
  });
  return remotes;
}

function parseUrls(remotesArray: unknown[], remotesString: string): GitRemote[] {
  const remotes: GitRemote[] = [];

  remotesArray.forEach((value, i) => {
    if (typeof value !== 'string') {
      throw `Malformed remotes provided: ${remotesString}`;
    }

    let name;
    if (i === 0) {
      name = 'origin';
    } else if (i === 1) {
      name = 'upstream';
    } else {
      name = `fork${i - 1}`;
    }
    remotes.push({ name, url: value });
  });
  return remotes;
}

export function configureProjectRemotes(
  devfile: devfileApi.Devfile,
  remotes: string,
  isDefaultDevfile: boolean,
): void {
  const parsedRemotes = getGitRemotes(remotes);

  // Determine the remote to set `checkoutFrom.remote` to
  let checkoutRemote = parsedRemotes.find(remote => remote.name === 'origin');
  if (!checkoutRemote) {
    checkoutRemote = parsedRemotes[0];
  }

  // Find the Git project in the devfile to configure remotes for
  let gitProject = getGitProjectForConfiguringRemotes(devfile.projects);
  if (gitProject === undefined) {
    if (!devfile.projects) {
      devfile.projects = [];
    }
    devfile.projects[0] = getProjectFromLocation(checkoutRemote.url, checkoutRemote.name);
    gitProject = devfile.projects[0].git;
  } else if (Object.keys(gitProject.remotes).includes('origin')) {
    checkoutRemote = { name: 'origin', url: gitProject.remotes.origin };
  }
  if (gitProject) {
    addRemotesToProject(gitProject, checkoutRemote, parsedRemotes);
  } else {
    console.warn('Failed to configure the project remotes.');
  }

  if (isDefaultDevfile) {
    const projectName = getProjectName(checkoutRemote.url);
    devfile.metadata.name = projectName;
    devfile.metadata.generateName = projectName;
  }
}

/**
 * Returns the Git project to replace remotes for
 */
function getGitProjectForConfiguringRemotes(projects: V230DevfileProjects[] | undefined) {
  if (!projects) {
    return undefined;
  }

  const gitProjects = projects.filter(project => project.git);
  if (gitProjects.length > 1) {
    throw new Error(
      'Configuring remotes is not supported when multiple Git projects found in Devfile.',
    );
  }

  if (gitProjects.length === 1) {
    return gitProjects[0].git;
  }

  return undefined;
}

/**
 * Add the remotes specified in `newRemotes` to the `gitProject`.
 * @param gitProject The Git project to add the new remotes to
 * @param checkoutRemote The Git remote to set checkoutFrom.remote to
 * @param newRemotes The array of new Git remotes to add
 */
function addRemotesToProject(
  gitProject: V230DevfileProjectsItemsGit,
  checkoutRemote: GitRemote,
  newRemotes: GitRemote[],
): void {
  const gitRemotes = newRemotes.reduce((map, remote) => {
    map[remote.name] = remote.url;
    return map;
  }, {});

  gitProject.remotes = Object.assign(gitProject.remotes, gitRemotes);
  gitProject.checkoutFrom = Object.assign(gitProject.checkoutFrom || {}, {
    remote: checkoutRemote.name,
  });
}
