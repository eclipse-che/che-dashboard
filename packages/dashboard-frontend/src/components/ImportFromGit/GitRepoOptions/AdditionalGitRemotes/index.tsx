/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  ActionGroup,
  Button,
  ButtonVariant,
  Form,
  FormFieldGroup,
  FormSection,
} from '@patternfly/react-core';
import { PlusCircleIcon } from '@patternfly/react-icons';
import { isEqual } from 'lodash';
import React from 'react';

import AdditionalGitRemote from '@/components/ImportFromGit/GitRepoOptions/AdditionalGitRemotes/gitRemote';
import styles from '@/components/ImportFromGit/GitRepoOptions/AdditionalGitRemotes/index.module.css';
import { GitRemote } from '@/components/WorkspaceProgress/CreatingSteps/Apply/Devfile/getGitRemotes';

export type Props = {
  onChange: (remotes: GitRemote[] | undefined, isValid: boolean) => void;
  remotes: GitRemote[] | undefined;
};
export type State = {
  remotes: GitRemote[];
};

export class AdditionalGitRemotes extends React.PureComponent<Props, State> {
  private readonly callbacks: { getValidation?: () => boolean }[];

  constructor(props: Props) {
    super(props);

    const remotes = this.props.remotes ? [...this.props.remotes] : [];
    if (remotes.length === 0) {
      remotes.push({ name: '', url: '' });
    }
    this.state = {
      remotes,
    };
    this.callbacks = remotes.map(() => ({}));
  }

  public componentDidMount(): void {
    const remotes = this.props.remotes || [];
    if (remotes.length > 0) {
      this.props.onChange(this.state.remotes, this.getValidate());
    }
  }

  public componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<State>): void {
    const remotes = this.props.remotes || [];
    const prevRemotes = prevProps.remotes || [];

    if (!isEqual(remotes, prevRemotes) && !isEqual(remotes, this.state.remotes)) {
      this.setState({
        remotes,
      });
    } else if (!isEqual(prevState.remotes, this.state.remotes)) {
      const validate = this.getValidate();
      this.props.onChange(this.state.remotes, validate);
    }
  }

  private getValidate(): boolean {
    const invalidIndex = this.callbacks.findIndex(
      callback => callback.getValidation && !callback.getValidation(),
    );
    return invalidIndex === -1;
  }

  private handleChange(index: number, remote: GitRemote): void {
    const remotes = [...this.state.remotes];
    remotes[index] = remote;

    this.setState({ remotes });
  }

  private handleAdd(): void {
    const remotes = [...this.state.remotes];
    remotes.push({ name: '', url: '' });

    this.setState({ remotes });
  }

  private handleRemove(index: number): void {
    const remotes = [...this.state.remotes];
    remotes.splice(index, 1);
    this.callbacks.splice(index, 1);

    this.setState({ remotes });
  }

  public buildGitRemotes(): React.JSX.Element {
    const { remotes } = this.state;
    this.callbacks.length = remotes.length;

    return (
      <>
        {remotes.map((remote: GitRemote, index: number) => {
          if (this.callbacks[index] === undefined) {
            this.callbacks[index] = {};
          }
          return (
            <AdditionalGitRemote
              key={'remote_' + index}
              remote={remote}
              onChange={(remote: GitRemote) => this.handleChange(index, remote)}
              callbacks={this.callbacks[index]}
              onDelete={() => this.handleRemove(index)}
            />
          );
        })}
      </>
    );
  }

  public render() {
    return (
      <Form isHorizontal={true}>
        <FormSection title="Additional Git Remotes" className={styles.gitRemotesTitle}>
          <FormFieldGroup className={styles.gitRemotesGroup}>
            {this.buildGitRemotes()}
            <ActionGroup style={{ marginTop: 0 }}>
              <Button
                onClick={() => this.handleAdd()}
                variant={ButtonVariant.link}
                style={{ textDecoration: 'none' }}
                isInline
                icon={<PlusCircleIcon />}
              >
                Add Remote
              </Button>
            </ActionGroup>
          </FormFieldGroup>
        </FormSection>
      </Form>
    );
  }
}
