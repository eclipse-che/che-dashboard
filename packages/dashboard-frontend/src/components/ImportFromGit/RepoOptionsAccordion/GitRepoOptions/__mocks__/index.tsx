/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions';

export class GitRepoOptions extends React.PureComponent<Props> {
  public render() {
    const { gitBranch, remotes, devfilePath, hasSupportedGitService, onChange } = this.props;

    return (
      <div>
        <div>Git Repo Options</div>
        <div data-testid="git-repo-options">{`${gitBranch}, ${JSON.stringify(remotes)}, ${devfilePath}, ${hasSupportedGitService}`}</div>
        <button
          onClick={() =>
            onChange(
              'newBranch',
              [{ name: 'test-updated', url: 'http://test' }],
              'newDevfilePath',
              true,
            )
          }
        >
          Git Repo Options Change
        </button>
      </div>
    );
  }
}
