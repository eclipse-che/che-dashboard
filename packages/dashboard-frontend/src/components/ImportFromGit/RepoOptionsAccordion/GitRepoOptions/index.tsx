/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import { isEqual } from 'lodash';
import React from 'react';

import { AdditionalGitRemotes } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/AdditionalGitRemotes';
import { GitBranchField } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/GitBranchField';
import { PathToDevfileField } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/PathToDevfileField';
import { GitRemote } from '@/components/WorkspaceProgress/CreatingSteps/Apply/Devfile/getGitRemotes';

export type Props = {
  gitBranch: string | undefined;
  remotes: GitRemote[] | undefined;
  devfilePath: string | undefined;
  hasSupportedGitService: boolean;
  onChange: (
    gitBranch: string | undefined,
    remotes: GitRemote[] | undefined,
    devfilePath: string | undefined,
    isValid: boolean,
  ) => void;
};

export type State = {
  gitBranch: string | undefined;
  remotes: GitRemote[] | undefined;
  devfilePath: string | undefined;
  isValid: boolean;
};

export class GitRepoOptions extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      gitBranch: props.gitBranch,
      remotes: props.remotes,
      devfilePath: props.devfilePath,
      isValid: true,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { gitBranch, remotes, devfilePath } = this.props;

    if (gitBranch !== prevProps.gitBranch) {
      this.setState({ gitBranch });
    }

    if (!isEqual(remotes, prevProps.remotes)) {
      this.setState({ remotes });
    }

    if (devfilePath !== prevProps.devfilePath) {
      this.setState({ devfilePath });
    }
  }
  private handleGitBranch(gitBranch: string | undefined) {
    const { remotes, devfilePath, isValid } = this.state;

    this.setState({ gitBranch });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }

  private handleRemotes(remotes: GitRemote[] | undefined, isValid: boolean) {
    const { gitBranch, devfilePath } = this.state;

    this.setState({ remotes, isValid });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }

  private handleDevfilePath(devfilePath: string | undefined) {
    const { gitBranch, remotes, isValid } = this.state;

    this.setState({ devfilePath });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }
  public render() {
    const { hasSupportedGitService } = this.props;
    const { gitBranch, remotes, devfilePath } = this.state;
    return (
      <Form isHorizontal={true} onSubmit={e => e.preventDefault()}>
        {hasSupportedGitService && (
          <GitBranchField
            onChange={gitBranch => this.handleGitBranch(gitBranch)}
            gitBranch={gitBranch}
          />
        )}
        <AdditionalGitRemotes
          onChange={(remotes: GitRemote[] | undefined, isValid: boolean) =>
            this.handleRemotes(remotes, isValid)
          }
          remotes={remotes}
        />
        <PathToDevfileField
          onChange={devfilePath => this.handleDevfilePath(devfilePath)}
          devfilePath={devfilePath}
        />
      </Form>
    );
  }
}
