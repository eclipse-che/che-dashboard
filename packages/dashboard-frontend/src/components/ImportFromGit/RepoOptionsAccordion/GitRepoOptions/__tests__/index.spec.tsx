/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import userEvent from '@testing-library/user-event';
import React from 'react';

import { GitRepoOptions } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions';
import { GitRemote } from '@/components/WorkspaceProgress/CreatingSteps/Apply/Devfile/getGitRemotes';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

jest.mock('@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/AdditionalGitRemotes');
jest.mock('@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/GitBranchField');
jest.mock('@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/PathToDevfileField');

const mockOnChange = jest.fn();

describe('GitRepoOptions', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot with default values', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('snapshot with all values', () => {
    const snapshot = createSnapshot(
      'test-git-branch',
      [{ name: 'test', url: 'http://test' }],
      'test-devfile-path',
    );
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('should remove "Git Branch" component when it is not supported', () => {
    const { reRenderComponent } = renderComponent(
      'test-git-branch',
      [{ name: 'test', url: 'http://test' }],
      'test-devfile-path',
    );

    expect(screen.queryByTestId('git-branch-component')).not.toBeNull();

    reRenderComponent(undefined, undefined, undefined, false);

    expect(screen.queryByTestId('git-branch-component')).toBeNull();
  });

  test('update Git Branch', () => {
    renderComponent('test-git-branch');

    const gitBranch = screen.getByTestId('git-branch');

    expect(gitBranch).toHaveTextContent('test-git-branch');

    const updateGitBranch = screen.getByRole('button', {
      name: 'Git Branch Change',
    });

    userEvent.click(updateGitBranch);

    expect(mockOnChange).toHaveBeenCalledWith('new-branch', undefined, undefined, true);
  });

  test('update Remotes', () => {
    renderComponent(undefined, [{ name: 'test', url: 'http://test' }]);

    const gitRemotes = screen.getByTestId('git-remotes');

    expect(gitRemotes).toHaveTextContent('[{"name":"test","url":"http://test"}]');

    const updateGitRemotes = screen.getByRole('button', {
      name: 'Git Remotes Change',
    });

    userEvent.click(updateGitRemotes);

    expect(mockOnChange).toHaveBeenCalledWith(
      undefined,
      [{ name: 'test-updated', url: 'http://test' }],
      undefined,
      true,
    );
  });

  test('update PathToDevfile', () => {
    renderComponent(undefined, undefined, 'test-devfile-path');

    const pathToDevfile = screen.getByTestId('devfile-path');

    expect(pathToDevfile).toHaveTextContent('test-devfile-path');

    const updatePathToDevfile = screen.getByRole('button', {
      name: 'Devfile Path Change',
    });

    userEvent.click(updatePathToDevfile);

    expect(mockOnChange).toHaveBeenCalledWith(undefined, undefined, 'new-devfile-path', true);
  });
});

function getComponent(
  gitBranch?: string | undefined,
  remotes?: GitRemote[] | undefined,
  devfilePath?: string | undefined,
  hasSupportedGitService: boolean = true,
) {
  return (
    <GitRepoOptions
      gitBranch={gitBranch}
      remotes={remotes}
      devfilePath={devfilePath}
      hasSupportedGitService={hasSupportedGitService}
      onChange={mockOnChange}
    />
  );
}
