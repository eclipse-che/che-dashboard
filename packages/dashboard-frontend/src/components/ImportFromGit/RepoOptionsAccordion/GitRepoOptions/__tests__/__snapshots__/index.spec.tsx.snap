// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitRepoOptions snapshot with all values 1`] = `
<form
  className="pf-c-form pf-m-horizontal"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    data-testid="git-branch-component"
  >
    <div>
      Git Branch
    </div>
    <div
      data-testid="git-branch"
    >
      test-git-branch
    </div>
    <button
      onClick={[Function]}
    >
      Git Branch Change
    </button>
  </div>
  <div>
    <div>
      Git Remotes
    </div>
    <div
      data-testid="git-remotes"
    >
      [{"name":"test","url":"http://test"}]
    </div>
    <button
      onClick={[Function]}
    >
      Git Remotes Change
    </button>
  </div>
  <div>
    <div>
      Devfile Path
    </div>
    <div
      data-testid="devfile-path"
    >
      test-devfile-path
    </div>
    <button
      onClick={[Function]}
    >
      Devfile Path Change
    </button>
  </div>
</form>
`;

exports[`GitRepoOptions snapshot with default values 1`] = `
<form
  className="pf-c-form pf-m-horizontal"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    data-testid="git-branch-component"
  >
    <div>
      Git Branch
    </div>
    <div
      data-testid="git-branch"
    />
    <button
      onClick={[Function]}
    >
      Git Branch Change
    </button>
  </div>
  <div>
    <div>
      Git Remotes
    </div>
    <div
      data-testid="git-remotes"
    >
      undefined
    </div>
    <button
      onClick={[Function]}
    >
      Git Remotes Change
    </button>
  </div>
  <div>
    <div>
      Devfile Path
    </div>
    <div
      data-testid="devfile-path"
    />
    <button
      onClick={[Function]}
    >
      Devfile Path Change
    </button>
  </div>
</form>
`;
